src/estimators/stratified_sketch.py

```diff
@@ -0,0 +1,242 @@
+# Copyright 2020 The Private Cardinality Estimation Framework Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import collections
+import copy
+import functools
+import sys
+import numpy as np
+from wfa_cardinality_estimation_evaluation_framework.estimators.base import SketchBase
+from wfa_cardinality_estimation_evaluation_framework.estimators.exact_set import ExactMultiSet
+
+ONE_PLUS = '1+'
+
+class StratifiedSketch(SketchBase):
+  """A frequency sketch that contains cardinality sketches per frequency bucket."""
+
+  @classmethod
+  def get_sketch_factory(cls):
+
+    def f(random_seed):
+      return cls(random_seed=random_seed)
+
+    return f
+  def __init__(self, max_freq, sketch_list, random_seed=None):
+    """Construct a Stratified sketch.
+
+    Args:
+      max_freq: the maximum targeting frequency level. For example, if
+        it is set to 3, then the sketches will include frequency=1, 2, and
+        frequency >= 3.
+      random_seed: This arg exists in order to conform to
+        simulator.EstimatorConfig.sketch_factory.
+      sketch_list: List of cardinality sketches of the same sketch type.
+
+    """
+    SketchBase.__init__(self)
+    # A dictionary that contains multiple sketches, which include:
+    # (1) sketches with frequency equal to k, where k < max_freq;
+    # (2) a sketch with frequency greater than or equal to max_freq;
+    # (3) a 1+ reach sketch.
+    self.sketches = {}
+    self.seed = random_seed
+    self.max_freq = max_freq
+
+  def init_from_exact_multi_set(self, exact_multi_set, CardinalitySketch=None):
```

Comment:
Several questions about the CardinalitySketch.

1. Is CardinalitySketch a class? How can we make sure that the sketch instance created from the CardinalitySketch are compatible, or namely using the same random seed and sketch parameters?

On the other hand, is it possible to use a sketch_factory so as to make sure that the sketches are compatible? For example,
```
def init_from_exact_multi_set(self, exact_multi_set, cardinality_sketch_factory):
  self.sketches[ONE_PLUS] = cardinality_sketch_factory(a_random_seed)
  for freq in range(1, self.max_freq + 1):
    self.sketches[str(freq)] = cardinality_sketch_factory(a_random_seed)
```

2. For either `cardinality_sketch_factory` or `CardinalitySketch`, do we want to make it an argument in the constructor (__init__)?

3. One more thing I am not sure is about the code style. When we use a class as an argument, do we use variable_name style or VariableName style?

Reply:
1. Per our conversation offline, changed it to a sketch factory.